lib/smart_answer/state_resolver.rb

```diff
@@ -0,0 +1,78 @@
+module SmartAnswer
+  class StateResolver
+    def initialize(flow)
+      @flow = flow
+    end
+
+    def state_from_response_store(response_store, requested_node = nil)
+      forwarding_responses = @flow.response_store == :session ? {} : response_store.all
+      state = start_state(forwarding_responses)
+      loop do
+        node_name = state.current_node.to_s
+        return state unless response_store.has?(node_name)
+
+        response = response_store.get(node_name)
+        return apply_response_to_state(state, response) if node_name == requested_node
+
+        new_state = transition_state_to_next_node(state, response)
+        return new_state if new_state.error
+        return new_state if @flow.node(new_state.current_node).outcome?
+
+        state = new_state
+      end
+    end
```

Comment:
Rather than inline comments which can get out of date, perhaps the methods need to be renamed? so rather than `transition_state_to_next_node` the method could be called `next_applicable_node_in_flow`?

Reply:
I've rewritten this code to be more explicit (which also gave me an opportunity to get rid of the loop which felt pretty clunky). Hopefully that's a further improvement.

Reply:
I've left comments on the new code - but overall I like the re-write, it's more succinct and probably does away for need for inline comments.